src/realm/object-store/impl/collection_notifier.cpp

```diff
@@ -51,14 +51,61 @@ CollectionNotifier::get_modification_checker(TransactionChangeInfo const& info,
         };
     }
 
-    if (m_related_tables.size() == 1) {
-        auto& object_set = info.tables.find(m_related_tables[0].table_key.value)->second;
+    // If the table in question has no outgoing links it will be the only entry in `m_related_tables`.
+    // In this case we do not need a `DeepChangeChecker` and check the modifications using the
+    // `ObjectChangeSet` within the `TransactionChangeInfo` for this table directly.
+    if (m_related_tables.size() == 1 && !all_callbacks_filtered()) {
+        auto root_table_key = m_related_tables[0].table_key;
+        auto& object_change_set = info.tables.find(root_table_key.value)->second;
         return [&](ObjectChangeSet::ObjectKeyType object_key) {
-            return object_set.modifications_contains(object_key);
+            return object_change_set.modifications_contains(object_key, {});
         };
     }
 
-    return DeepChangeChecker(info, *root_table, m_related_tables);
+    if (all_callbacks_filtered()) {
+        return KeyPathChangeChecker(info, *root_table, m_related_tables, m_key_path_arrays);
+    }
+    else if (any_callbacks_filtered()) {
+        // In case we have some callbacks, we need to combine the unfiltered `DeepChangeChecker` with
+        // the filtered `KeyPathChangeChecker` to make sure we send all expected notifications.
+        return [&info, root_table, this](ObjectChangeSet::ObjectKeyType object_key) {
+            auto key_path_change_checker =
+                KeyPathChangeChecker(info, *root_table, m_related_tables, m_key_path_arrays);
+            auto deep_change_checker = DeepChangeChecker(info, *root_table, m_related_tables, m_key_path_arrays);
```

Comment:
These two should be constructed outside of the lambda and captured by it so that they're constructed once rather than once per key checked.

Reply:
@DominicFrei I addressed this in 34d7c90. Could be a bit tricky to get right.

Reply:
Thank you, @jedelbo ! I've merged your PR. 💪